enhancements/agent-installer/automated-workflow-for-agent-based-installer.md

```diff
@@ -0,0 +1,624 @@
+independently of the installer, potentially breaking compatibility
+between the 2 tools. The risk of this happening is low, because the
+API is a relatively simple wrapper for nmstate config content that is
+not parsed directly by ZTP. If the API is changed, the new version
+will need to be backwards-compatible for ZTP users, which should allow
+us to bring the installer up to date without significant impact.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. What do we do with the installation logs?
+
+   - Write them to disk on each host in a way that allows someone
+     logging in to find them or allows a Pod to mount the directory
+     they are in and `cat` them to be collected via the OpenShift API.
+   - Expect the orchestration tool to download them before allowing
+     deployment to continue.
+   - Can we take advantage of or replicate the installer's [bootstrap
+     service
+     records](https://github.com/openshift/installer/blob/6d778f911e79afad8ba2ff4301eda5b5cf4d8e9e/docs/dev/bootstrap_services.md)
+     behavior?
+
+2. Is the workflow different for single-node? Do we need to start the
+   service and go through the orchestration steps for one node? If we
+   don't, how do we handle the validations?
+3. How do we implement a feature flag in the installer?
+
+    - A build flag, like what we do for ARO and OKD today
+    - An environment variable that enables the commands under the
+      `agent` prefix, which are otherwise completely disabled and
+      inactive
+    - Hide the commands under the `agent` prefix from the `--help`
+      output, but otherwise leave them fully active and operational
+
+### Test Plan
+
+We will implement CI jobs to verify deployment of clusters in all
+three topologies (single-node, compact, and HA) using VMs simulating
+bare metal hosts. We will configure the jobs so that they run for
+changes in the dev-scripts, installer, assisted-service, and
+assisted-agent repositories to ensure that regressions are not
+introduced.
+
+We will implement end-to-end CI jobs to exercise the
+parallel/serial/upgrade suites running periodically as release
+informers/blockers to ensure that clusters created by the agent-based
+installer meet the expectations of all OpenShift clusters.
+
+We will use unit and functional tests on the installer repository to
+test the image generator using different types of inputs.
+
+### Graduation Criteria
+
+We will use a fork or branch of the installer to prototype this work,
+and then merge it into the installer repository protected by a feature
+flag. The minimum viable version of the new feature will be exposed as
+tech preview (deployment failures may not receive immediate attention,
+but clusters successfully deployed using the tool will be fully
+supported).
+
+#### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- The minimum viable version of the new installer may only work in
```

Comment:
It's not clear if this means that it must work in disconnected to go Tech Preview, or something else.

Reply:
The first item in the TP->GA list below is to work in disconnected settings. This was trying to express that we might not necessarily get to that feature in the first iteration. Should it be rephrased, or is the statement incorrect?

Reply:
I guess we could download the RHCOS ISO from the internet (the installer already knows where it is) and implement extracting it from the payload. So that's one task we could delay.

Reply:
Do we know that all of the mirror registry settings can be passed through the various API layers to end up being used correctly? It's fine if it works, I'm just trying to set expectations correctly here.

Reply:
I think if they're not then it won't work at all - nothing magically knows how to find a release image except the install-config defaults. But the RHCOS extraction is a reason to not require disconnected support for TP. I do still expect we'll probably get there anyway.